File: python/dijkastra.py
import sys

def prin(v,key_val):
	print("Vertex\tDistance")
	for i in range(1,v):
		print(str(i)+"\tWeight: "+str(key_val[i]))
	

def mini(key_val,processed,v):
	mi_val=sys.maxsize
	mi_index=-1
	for i in range(v):
		if(key_val[i]<=mi_val and processed[i]==0):
			mi_val=key_val[i]
			mi_index=i
		
	
	return mi_index


def mst(weights,v,o):
	key_val=[]
	processed=[]

	for i in range(v):
		key_val.append(sys.maxsize)
		processed.append(0)
		
	key_val[o]=0
	
	for i in range(v-1):
		u=mini(key_val,processed,v)

		processed[u]=1
		for j in range(v):
			if(weights[u][j]!=0 and processed[j]==0 and weights[u][j]+key_val[u]<key_val[j]):
				key_val[j]= weights[u][j]+key_val[u]
			
		
	
	prin(v,key_val)


print("Enter the no. of vertices:\n")
v=int(input())

weights=[]
for i in range(v):
	w=[]
	for j in range(v):
		print("Enter the weight of between "+str(i+1)+"-"+str(j+1)+" vertices :")
		w.append(int(input()))
		if(w[j]==-1):
			w[j]=sys.maxsize
	weights.append(w)
mst(weights,v,0)

File: java/dijkastra.java
import java.util.*;

class dijkastra{

	public static int min(int key_val[],int processed[],int v){
		int mi_val=Integer.MAX_VALUE,mi_index=-1;
		for(int i=0;i<v;i++){
			if(key_val[i]<=mi_val && processed[i]==0){
				mi_val=key_val[i];
				mi_index=i;
			}
		}
		return mi_index;
	}

	public static void mst(int weights[][],int v,int o){
		int key_val[]=new int[v];
		int processed[]=new int[v];

		for(int i=0;i<v;i++){
			key_val[i]=Integer.MAX_VALUE;
			processed[i]=0;
		}

		key_val[o]=0;
		for(int i=0;i<v-1;i++){
			int u=min(key_val,processed,v);

			processed[u]=1;
			for(int j=0;j<v;j++){
				if(weights[u][j]!=0 && processed[j]==0 && weights[u][j]+key_val[u]<key_val[j]){
					key_val[j]= weights[u][j]+key_val[u];
				}
			}
		}

		print(v,key_val);
	}

	public static void print(int v,int key_val[]){
		System.out.println("Vertex\tDistance");
		for(int i=1;i<v;i++){
			System.out.println(i+"\tWeight: "+key_val[i]);
		}
	}

	public static void main(String args[]){

		Scanner sc=new Scanner(System.in);
		System.out.println("Enter the no. of vertices:");
		int v=sc.nextInt();

		int weights[][]=new int[v][v];

		for(int i=0;i<v;i++){
			for(int j=0;j<v;j++){
				System.out.println("Enter the weight of between "+(i+1)+"-"+(j+1)+" vertices:");
				weights[i][j]=sc.nextInt();
			}
		}

		mst(weights,v,0);
	}
}